// jest-preset-angular/build/presets/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultEsmPreset = exports.defaultPreset = void 0;
const tslib_1 = require("tslib");
const serializers_1 = tslib_1.__importDefault(require("../serializers"));
const baseConfig = {
    globals: {
        'ts-jest': {
            tsconfig: '<rootDir>/tsconfig.spec.json',
            stringifyContentPathRegex: '\\.(html|svg)$',
        },
    },
    testEnvironment: 'jsdom',
    moduleFileExtensions: ['ts', 'html', 'js', 'json', 'mjs'],
    snapshotSerializers: serializers_1.default,
};
const defaultPreset = Object.assign(Object.assign({}, baseConfig), { transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'], transform: {
        '^.+\\.(ts|js|mjs|html|svg)$': 'jest-preset-angular',
    } });
exports.defaultPreset = defaultPreset;
const defaultEsmPreset = Object.assign(Object.assign({}, baseConfig), { extensionsToTreatAsEsm: ['.ts'], globals: {
        'ts-jest': Object.assign(Object.assign({}, baseConfig.globals['ts-jest']), { useESM: true }),
    }, moduleNameMapper: {
        tslib: 'tslib/tslib.es6.js',
    }, transform: {
        '^.+\\.(ts|js|html|svg)$': 'jest-preset-angular',
    }, transformIgnorePatterns: ['node_modules/(?!tslib)'] });
exports.defaultEsmPreset = defaultEsmPreset;
